plugins/optimization-detective/storage/class-od-url-metric-stored-context.php

```diff
@@ -0,0 +1,80 @@
+<?php
+/**
+ * Optimization Detective: OD_URL_Metric_Stored_Context class
+ *
+ * @package optimization-detective
+ * @since n.e.x.t
+ */
+
+// Exit if accessed directly.
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * Context for when a URL metric is successfully stored via the REST API.
+ *
+ * @since n.e.x.t
+ * @access private
+ */
+final class OD_URL_Metric_Stored_Context {
+
+	/**
+	 * Request.
+	 *
+	 * @var WP_REST_Request<array<string, mixed>>
+	 * @readonly
+	 */
+	public $request;
+
+	/**
+	 * ID for the URL metric post.
+	 *
+	 * @var int
+	 * @readonly
+	 */
+	public $post_id;
+
+	/**
+	 * URL metric group collection.
+	 *
+	 * @var OD_URL_Metric_Group_Collection
+	 * @readonly
+	 */
+	public $url_metric_group_collection;
+
+	/**
+	 * URL metric group.
+	 *
+	 * @var OD_URL_Metric_Group
+	 * @readonly
+	 */
+	public $url_metric_group;
+
+	/**
+	 * URL metric.
+	 *
+	 * @var OD_URL_Metric
+	 * @readonly
+	 */
+	public $url_metric;
+
+	/**
+	 * Constructor.
+	 *
+	 * @phpstan-param WP_REST_Request<array<string, mixed>> $request
+	 *
+	 * @param WP_REST_Request                $request                     REST API request.
+	 * @param int                            $post_id                     ID for the URL metric post.
+	 * @param OD_URL_Metric_Group_Collection $url_metric_group_collection URL metric group collection.
+	 * @param OD_URL_Metric_Group            $url_metric_group            URL metric group.
+	 * @param OD_URL_Metric                  $url_metric                  URL metric.
+	 */
+	public function __construct( WP_REST_Request $request, int $post_id, OD_URL_Metric_Group_Collection $url_metric_group_collection, OD_URL_Metric_Group $url_metric_group, OD_URL_Metric $url_metric ) {
```

Comment:
This class has a terrible amount of redundancy. However, in newer versions of PHP this can be made much simpler, thanks to [Constructor Promotion](https://www.php.net/manual/en/language.oop5.decon.php#language.oop5.decon.constructor.promotion).

So this could be replaced with:

```php
final class OD_URL_Metric_Stored_Context {
	public function __construct(
		public WP_REST_Request $request,
		public int $post_id,
		public OD_URL_Metric_Group_Collection $url_metric_group_collection,
		public OD_URL_Metric_Group $url_metric_group,
		public OD_URL_Metric $url_metric
	) {}
}
```

This works in PHP 8.0+: https://3v4l.org/OqvTC

So we can't take advantage of it. But it will greatly simplify this class and `OD_Tag_Visitor_Context`.

Reply:
How do you mean redundancy? Just because the properties are declared in several places?

Reply:
Yes. There are 4 instances of the `$url_metric_group_collection` string in this class file when there could just be 1.

Reply:
But that's just about how PHP (and many other languages work) right? I wouldn't necessarily call that redundancy, it's just more code to write. :)

Reply:
Correct. The verbosity can be reduced in the future.